src/pool/srv_pool.c

```diff
@@ -4266,6 +4414,180 @@ ds_pool_update(uuid_t pool_uuid, crt_opcode_t opc,
+		int32_t *domains, bool *updated_p, uint32_t *map_version_p,
+		struct rsvc_hint *hint)
+{
+	struct pool_buf		*map_buf;
+	struct pool_map		*map = NULL;
+	uint32_t		map_version;
+	uuid_t			*uuids = NULL;
+	bool			updated = false;
+	int			ntargets;
+	int			rc;
+
+	ntargets = nnodes * dss_tgt_nr;
+
+	/* Create a temporary pool map based on the last committed version. */
+	rc = read_map(tx, &svc->ps_root, &map);
+	if (rc != 0)
+		return rc;
+
+	map_version = pool_map_get_version(map) + 1;
+
+	/* Prepare the pool map attribute buffers. */
+	map_buf = pool_buf_alloc(ndomains + nnodes + ntargets);
+	if (map_buf == NULL)
+		return -DER_NOMEM;
+
+	/* Make a sorted target UUID array to determine target IDs. */
+	D_ALLOC_ARRAY(uuids, nnodes);
+	if (uuids == NULL)
+		D_GOTO(out_map_buf, rc = -DER_NOMEM);
+	memcpy(uuids, target_uuids, sizeof(uuid_t) * nnodes);
+	qsort(uuids, nnodes, sizeof(uuid_t), uuid_compare_cb);
+
+	/* Fill the pool_buf out. */
+	rc = pool_buf_fill_extend(map, map_buf, nnodes, map_version, ndomains,
+			    domains, target_uuids, rank_list, uuids, &updated);
+	if (rc != 0)
+		D_GOTO(out_uuids, rc);
+	if (!updated)
+		D_GOTO(out_map_buf, rc);
+
+	/* Extend the current pool map */
+	rc = pool_map_extend(map, map_version, map_buf);
+	if (rc != 0)
+		D_GOTO(out_map, rc);
+
+	/* Write the new pool map. */
+	rc = pool_buf_extract(map, &map_buf);
+	if (rc != 0)
+		D_GOTO(out_map, rc);
+
+	rc = write_map_buf(tx, &svc->ps_root, map_buf, map_version);
+	if (rc != 0)
+		D_GOTO(out_map, rc);
+
+	rc = rdb_tx_commit(tx);
+	if (rc != 0) {
+		D_DEBUG(DB_MD, DF_UUID": failed to commit: "DF_RC"\n",
+			DP_UUID(svc->ps_uuid), DP_RC(rc));
+			D_GOTO(out_map, rc);
+	}
+
+	updated = true;
+
+	/* Update svc->ps_pool to match the new pool map. */
+	rc = ds_pool_tgt_map_update(svc->ps_pool, map_buf, map_version);
+	if (rc != 0) {
+		/*
+		* We must resign to avoid handling future requests with a
+		* stale pool map cache.
+		*/
+		rdb_resign(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term);
+		rc = 0;
+		goto out_map;
+	}
+
+	ds_rsvc_request_map_dist(&svc->ps_rsvc);
+
+out_map:
+	if (map_version_p != NULL)
+		*map_version_p = pool_map_get_version((map == NULL || rc != 0) ?
+						      svc->ps_pool->sp_map :
+						      map);
+	ABT_rwlock_unlock(svc->ps_lock);
+	rdb_tx_end(tx);
+
+out_uuids:
+	D_FREE(uuids);
+
+out_map_buf:
+	pool_buf_free(map_buf);
+
+	if (updated_p)
+		*updated_p = updated;
+	if (map)
+		pool_map_decref(map);
+
+	return rc;
+}
+
+int
+ds_pool_extend(uuid_t pool_uuid, struct rsvc_hint *hint, uint32_t nnodes,
+		uuid_t target_uuids[], d_rank_list_t *rank_list,
+		uint32_t ndomains, int32_t *domains, uint32_t *map_version_p)
+
+{
+	struct pool_svc		*svc;
+	struct rdb_tx		tx;
+	bool			updated;
+	int rc;
+
+	rc = pool_svc_lookup_leader(pool_uuid, &svc, hint);
+	if (rc != 0)
+		D_GOTO(out_svc, rc);
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0)
+		D_GOTO(out_svc, rc);
+	ABT_rwlock_wrlock(svc->ps_lock);
```

Comment:
Please do not put lock/unlock in two different functions. Very easy to cause bugs.

Reply:
see above